graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/FederationDirectives.java

```diff
@@ -28,136 +29,89 @@
 public final class FederationDirectives {
     /* Directive locations */
 
+    public static final String keyName = "key";
+    public static final String externalName = "external";
+    public static final GraphQLDirective external = newDirective()
+            .name(externalName)
+            .validLocations(FIELD_DEFINITION)
+            .build();
+
+    /* fields: _FieldSet */
+    public static final String requiresName = "requires";
+    public static final String providesName = "provides";
+    public static final String extendsName = "extends";
+
+    /* directive @key(fields: _FieldSet!) on OBJECT | INTERFACE */
+    public static final GraphQLDirective extends_ = newDirective()
+            .name(extendsName)
+            .validLocations(OBJECT)
+            .build();
+    public static final Set<GraphQLDirective> allDirectives = new HashSet<>();
+    public static final Set<SDLDefinition> allDefinitions = new HashSet<>();
     private static final DirectiveLocation DL_OBJECT = newDirectiveLocation()
             .name("OBJECT")
             .build();
 
+    /* directive @external on FIELD_DEFINITION */
+    public static final DirectiveDefinition extendsDefinition = newDirectiveDefinition()
+            .name(extendsName)
+            .directiveLocations(Arrays.asList(DL_OBJECT))
+            .build();
     private static final DirectiveLocation DL_INTERFACE = newDirectiveLocation()
             .name("INTERFACE")
             .build();
-
     private static final DirectiveLocation DL_FIELD_DEFINITION = newDirectiveLocation()
             .name("FIELD_DEFINITION")
             .build();
 
-    /* fields: _FieldSet */
-
-    private static final GraphQLArgument fieldsArgument = newArgument()
-            .name("fields")
-            .type(new GraphQLNonNull(_FieldSet.type))
+    /* directive @requires(fields: _FieldSet!) on FIELD_DEFINITION */
+    public static final DirectiveDefinition externalDefinition = newDirectiveDefinition()
+            .name(externalName)
+            .directiveLocations(Arrays.asList(DL_FIELD_DEFINITION))
             .build();
-
-    private static final GraphQLArgument fieldsArgument(String value) {
-        return newArgument(fieldsArgument)
-                .value(value)
-                .build();
-    }
-
-    private static final InputValueDefinition fieldsDefinition = newInputValueDefinition()
+    private static final GraphQLArgument fieldsArgument = newArgument()
             .name("fields")
-            .type(new NonNullType(new TypeName(_FieldSet.typeName)))
+            .type(GraphQLNonNull.nonNull(new GraphQLTypeReference(_FieldSet.typeName)))
```

Comment:
I know this can sometimes have some strange behavior in relying on reflection, and though I think this is fine, I'm always a little cautious in relying on it if we don't have to, and I'm not sure why we need this to be a type reference rather than the type itself?

Reply:
Funnily enough because `graphql.language.InputValueDefinition.Builder#type` wants nothing to do with the `graphql.schema` concept of a type.

Those 2 concepts are only connected through `Object`:

![Type](https://user-images.githubusercontent.com/8641/60950867-d39f0700-a2ac-11e9-8992-beb7a60b46c2.png)

Reply:
sad times.